microservices/download/utils.js

```diff
@@ -1,14 +1,44 @@
 const superagent = require("superagent");
 
 function query_os(req) {
-  let prov = req.query.os ?? "";
+  let raw = req; // The URL string containing any number of query params.
+  let prov = undefined;
+
+  let full = raw.split("&");
+
+  for (let i = 0; i < full.length; i++) {
+    if (full[i].startsWith("os=")) {
+      prov = full[i].split("=")[1];
+      break;
+    }
+  }
+
+  if (prov === undefined) {
+    return false;
+  }
+
```

Comment:
Did the format of `req` change from removing `express`?

The previous `let prov = req.query.os ?? "";` was obviously fewer lines than this, but I admit the new code isn't actually hard to read. Just longer than "check if input is null, else use an empty string". So I assume there is no `req.query.os` anymore to look at.

Reply:
I don't know if `for...of` would be worth using here, as long as it works it works. And maybe the above is faster, IDK.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

`for...of` is a bit of a simpler way to iterate through an array, I tend to like `for...of`, from the perspective of an author, at least. Not an expert enough to say if there is like a performance argument to be had, or what.

Reply:
Best I can do to simplify + subjectively more intuitive variable names is:

```js
function query_os(query_string) {
  // query_string should be an "&"-separated list of any number of query params

  let parsed_os = undefined;

  const params = query_string.split("&")

  for (const param of params) {
    if (param.startsWith("os=")) {
      parsed_os = param.split("=")[1];
      break;
    }
  }

  let valid = [ "linux", "arm_linux", "silicon_mac", "intel_mac", "windows" ];

  return valid.includes(parsed_os) ? parsed_os : false;
}
```

_(I know the JS convention is camelCase, but I saw underscores and went with it.)_

Dropping this in place works (on my machine, anyway).

If you like it, you can have it and/or I will push it to the branch for you if requested, but the existing version in the PR is good as-is anyhow, no worries.

ALSO: I volunteer to go ahead and fix the merge conflict real quick if you want, just let me know and I'll do it.

(Felt a bit bad holding this up with review, so whatever I can do to get it across the finish line. 👍 )